### PR TITLE
GraphQL - Add stats for a user (profile)

### PR DESCRIPTION
**Pull request by toyhammered**
# What
Add the anime/manga_amount_consumed and anime/manga_category_breakdown to a user.

# Checklist

- [ ] All files pass Rubocop
- [ ] Any complex logic is commented
- [ ] Any new systems have thorough documentation
- [ ] Any user-facing changes are behind a feature flag (or: explain why they can't be)

- [ ] All the tests pass
- [ ] Tests have been added to cover the new code

### CODE REVIEW

File: app/graphql/types/stat.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  description 'The basic stat information.'
+
+  field :id, ID, null: false
+
+  field :profile, Types::Profile,
+    null: false,
+    description: 'The profile related to the user for this stat.',
+    method: :user
+
+  field :data, GraphQL::Types::JSON,
```
Comment (NuckChorris) [inline app/graphql/types/stat.rb]: Big JSON blobs are bad GraphQL practice, so I think we should create separate types defining the schema for each stat type and implementing a Stat interface like we do for Media

File: app/graphql/types/stat_type.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+class Types::StatType < Types::BaseObject
```
Comment (NuckChorris) [inline app/graphql/types/stat_type.rb]: I don't think this is a good name, what if you renamed this to ProfileStats to be more clear? 🤔

**Review by NuckChorris — CHANGES_REQUESTED** [review]

File: app/graphql/types/stat.rb (hunk 1 above)
Comment (toyhammered) [reply app/graphql/types/stat.rb]: fineeeeeeeeeee. lol

File: app/graphql/types/profile_stats/anime_amount_consumed.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Types::ProfileStats::AnimeAmountConsumed < Types::BaseObject
+  implements Types::ProfileStats::ProfileStatInterface
+
+  field :data, Types::ProfileStats::AmountConsumed,
+    null: false,
+    description: 'The breakdown of this specific stat.',
+    method: :stats_data
```
Comment (NuckChorris) [inline app/graphql/types/profile_stats/anime_amount_consumed.rb]: Is there a reason for this, versus using an interface like `implements Types::ProfileStats::AmountConsumed`?

File: app/graphql/types/profile_stats/amount_consumed.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+class Types::ProfileStats::AmountConsumed < Types::BaseObject
+  description 'Generic Amount Consumed based on Media'
+
+  field :time, Integer,
+    null: false,
+    description: 'Total time spent in minutes.'
+
+  def time
+    object['time']
+  end
```
Comment (NuckChorris) [inline app/graphql/types/profile_stats/amount_consumed.rb]: This should probably only be on the AnimeAmountConsumed

File: app/graphql/types/profile_stats/amount_consumed.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    null: false,
+    description: 'Total amount of media.'
+
+  def media
+    object['media']
+  end
+
+  field :units, Integer,
+    null: false,
+    description: 'Total progress per media including reconsuming.'
```
Comment (NuckChorris) [inline app/graphql/types/profile_stats/amount_consumed.rb]: "per media" seems confusing.  How about "total progress of library including reconsuming"?

File: app/graphql/types/profile_stats/category_breakdown.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  def total
+    object['total']
+  end
+
+  field :categories, Types::Map,
+    null: false,
+    description:
+      <<~DESCRIPTION.squish
+        A Dictionary of category_id -> count for all categories
```
Comment (NuckChorris) [inline app/graphql/types/profile_stats/category_breakdown.rb]: Call it a Map since that's the type name, we should be consistent ;)

**Review by NuckChorris — CHANGES_REQUESTED** [review]

File: app/graphql/types/profile_stats/anime_amount_consumed.rb (hunk 3 above)
Comment (toyhammered) [reply app/graphql/types/profile_stats/anime_amount_consumed.rb]: 🤔 no, interface is most def the right way to go.

**Review by NuckChorris — APPROVED** [review]

**Review by wopian — APPROVED** [review]

**Pull request merged**